src/cascadia/TerminalSettingsEditor/NewTabMenuViewModel.cpp

```diff
@@ -0,0 +1,765 @@
+        {
+            _CurrentFolder.Inlining(value);
+            _NotifyChanges(L"CurrentFolderInlining");
+        }
+    }
+
+    bool NewTabMenuViewModel::CurrentFolderAllowEmpty() const
+    {
+        if (!_CurrentFolder)
+        {
+            return {};
+        }
+        return _CurrentFolder.AllowEmpty();
+    }
+
+    void NewTabMenuViewModel::CurrentFolderAllowEmpty(bool value)
+    {
+        if (_CurrentFolder && _CurrentFolder.AllowEmpty() != value)
+        {
+            _CurrentFolder.AllowEmpty(value);
+            _NotifyChanges(L"CurrentFolderAllowEmpty");
+        }
+    }
+
+    Windows::Foundation::Collections::IObservableVector<Editor::NewTabMenuEntryViewModel> NewTabMenuViewModel::CurrentView() const
+    {
+        if (!_CurrentFolder)
+        {
+            return _rootEntries;
+        }
+        return _CurrentFolder.Entries();
+    }
+
+    void NewTabMenuViewModel::UpdateSettings(const Model::CascadiaSettings& settings)
+    {
+        _Settings = settings;
+        _NotifyChanges(L"AvailableProfiles");
+
+        SelectedProfile(AvailableProfiles().GetAt(0));
+
+        _rootEntries = _ConvertToViewModelEntries(_Settings.GlobalSettings().NewTabMenu(), _Settings);
+
+        _rootEntriesChangedRevoker = _rootEntries.VectorChanged(winrt::auto_revoke, [this](auto&&, const IVectorChangedEventArgs& args) {
+            switch (args.CollectionChange())
+            {
+            case CollectionChange::Reset:
+            {
+                // fully replace settings model with view model structure
+                std::vector<Model::NewTabMenuEntry> modelEntries;
+                for (const auto& entry : _rootEntries)
+                {
+                    modelEntries.push_back(NewTabMenuEntryViewModel::GetModel(entry));
+                }
+                _Settings.GlobalSettings().NewTabMenu(single_threaded_vector<Model::NewTabMenuEntry>(std::move(modelEntries)));
+                return;
+            }
+            case CollectionChange::ItemInserted:
+            {
+                const auto& insertedEntryVM = _rootEntries.GetAt(args.Index());
+                const auto& insertedEntry = NewTabMenuEntryViewModel::GetModel(insertedEntryVM);
+                _Settings.GlobalSettings().NewTabMenu().InsertAt(args.Index(), insertedEntry);
+                return;
+            }
+            case CollectionChange::ItemRemoved:
+            {
+                _Settings.GlobalSettings().NewTabMenu().RemoveAt(args.Index());
+                return;
+            }
+            case CollectionChange::ItemChanged:
+            {
+                const auto& modifiedEntry = _rootEntries.GetAt(args.Index());
+                _Settings.GlobalSettings().NewTabMenu().SetAt(args.Index(), NewTabMenuEntryViewModel::GetModel(modifiedEntry));
+                return;
+            }
+            }
+        });
+
+        // Clear CurrentFolder to reset the view
+        _CurrentFolder = nullptr;

```

Comment:
This one took a while to fix. Since we have a PropertyChanged handler for "CurrentFolder" that navigates to the right subpage and updates the breadcrumbs, whenever we update CurrentFolder, there's a risk of the state being messed up. So, for most of this PR, I made it so that we'd just always navigate to the main page (to avoid doing the work of finding the folder and building the breadcrumbs).

Went ahead and did it anyways in [aeab125](https://github.com/microsoft/terminal/pull/18015/commits/aeab125e44a2bbb42974b5163f0c700dfbf5e87f). Now, we _do_ navigate to the correct folder if it exists by matching the name. If no match was found, we just go back to the main page.